open_instruct/rejection_sampling_multiple_rewards.py

```diff
@@ -0,0 +1,321 @@
+        remove_columns=ds.column_names
+    )
+
+    ### So this code handles only classification, I should also handle other models judges like Llama3
+    model = AutoModelForSequenceClassification.from_pretrained(
+        model_name_or_path,
+        torch_dtype=torch.bfloat16,
+        attn_implementation="flash_attention_2",
+    )
+    model = model.to(device)
+
+    # Initialize a data collator to handle dynamic padding of input sequences
+    data_collator = DataCollatorWithPadding(tokenizer=tokenizer)
+    current_batch_size = args.max_forward_batch_size
+    # NOTE: two optimizations here:
+    # 1. we sort by input_ids length to reduce padding at first
+    # 2. we shrink the batch size if we run out of memory (so initially we can use a large batch size)
+    input_ids_lengths = [len(x) for x in ds["input_ids"]] #  input_ids_lengths: (num_items_in_shard,)
+
+    sorted_indices = np.argsort(input_ids_lengths) # Get indices that would sort the input lengths
+
+    # Initialize a list to store the scores for each item in the shard
+    scores = []
+    i = 0
+    while i < len(ds):
+        with torch.no_grad():
+            data = ds[sorted_indices[i:i + current_batch_size]]
+            try:
+                input_ids = data_collator(data)["input_ids"].to(device)
+                _, score, _ = get_reward(model, input_ids, tokenizer.pad_token_id, 0)
+                # score = (batch_size, )
+                scores.extend(score.cpu().tolist()) # convert the tensor score to a list
+                i += current_batch_size
+                print(f"processing: {i}:{i + current_batch_size}/{len(ds)}")
+            except torch.cuda.OutOfMemoryError:
+                if current_batch_size == 1:
+                    raise ValueError("Out of memory even with batch size 1")
+                current_batch_size //= 2
+                print(f"Reducing batch size to {current_batch_size}")
+                continue
+    # restore the original order
+    scores = np.array(scores)
+    scores = scores[np.argsort(sorted_indices)]
+    return torch.tensor(scores)
+
+
+def majority_vote(offsets_per_model: dict[str, torch.tensor]) -> torch.tensor:
+    """
+    offsets_per_model: offsets returned by each model. each tensor is of shape (n_prompts,) indicating best/worst completion offset per prompt
+    """
+    # Determine the number of samples
+    num_samples = offsets_per_model[next(iter(offsets_per_model))].size(0)
+    # Initialize tensor to store the majority votes
+    majority_votes = torch.zeros(num_samples, dtype=torch.long)
+
+    # Tally the votes and determine the majority vote for each sample
+    for i in range(num_samples):
+        # Collect votes from all models for the current sample
+        votes = [offsets_per_model[model][i].item() for model in offsets_per_model]
+        # Determine the most common vote
+        majority_vote = Counter(votes).most_common(1)[0][0]
```

Comment:
```suggestion
        counter = Counter(votes)
        # Try to get ther majority vote, but if all models disagree, we randomly choose one
        if len(offsets_per_model) != len(counter):
            majority_vote = counter.most_common(1)[0][0]
        else:
            majority_vote = votes[np.random.randint(len(votes))]
```

One special case: what if the models disagree all the time? In that case we should randomly choose one. The current implementation I think might have always choose the the first RM.

Reply:
Good catch, will fix this after the meeting.